File: solutions/clojure/pascals-triangle/2/src/pascals_triangle.clj
(ns pascals-triangle)


(defn calc-next-row [last-row]
  "right and left of the previous row"
  (concat [1N]
          (->> last-row
               (partition 2 1)
               (map (partial apply +)))
          [1N]))


(def triangle (iterate calc-next-row [1]))

(defn row [n]
  (->> triangle
       (drop (dec n))
       first))

File: solutions/scala/pascals-triangle/2/src/main/scala/PascalsTriangle.scala
object PascalsTriangle {
  def rows(count: Int): List[List[Int]] =
    if count <= 0 then List()
    else
      (1 until count).foldLeft(List(List(1))) { (triangle, _) =>
        triangle :+ nextRow(triangle.last)
      }

  private def nextRow(row: List[Int]): List[Int] =
    (0 :: row).zip(row :+ 0).map(_ + _)
}
